LanguageFeatures/Subtyping/dynamic/test_types/function_type_covariant_arg_A01.dart

```diff
@@ -0,0 +1,31 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion We say that a type T0 is a subtype of a type T1 (written
+/// T0 <: T1) when:
+/// Function Type/Function: T0 is a function type and T1 is Function
+///
+/// @description Check that if type `T0` is a function type with a covariant
+/// arguments and `T1` is a function type with arguments that are supertypes of
+/// `T0` arguments, then `T0` is a subtype of `T1`.
```

Comment:
We can't really say that. First, this might seem like a tiny thing, but the specification documents do use the word 'parameter' or the phrase 'formal parameter' to denote the declaration that introduces support for receiving values at the start of a method/setter invocation, and they use 'argument' or 'actual argument' to denote the actual value which is being passed. So we should say '`T0` is a function type with a formal parameter that [something]'.

Next, function types can never have a parameter which is covariant (covariant-by-declaration or covariant-by-class), function types just have formal parameters with types. It's entirely a property of a formal parameter of a method/setter declaration that one or more parameters are covariant(-by-whatever).

This also implies that there is no need to test the treatment of function types with covariant parameter types, because they don't exist.

However, a _member signature_ can have formal parameters which are covariant. When we're checking for correct override relations, we transform the member signature into a function type, and in this case we're transforming the type of each parameter which is covariant-by-declaration into a top type (that's `Object?` today) before we check that the overriding declaration's function type is a subtype of the overridden declaration's function type.

Perhaps this is what these tests do?

In any case, the static type of a tear-off of `foo` in `C` is `void Function(int, [String])`, and the static type of `foo` in `A` is `void Function(num, [Object])`, but in this case we only use the latter, so lines 24 and 25 are OK at compile time.

The run-time type of that same tear-off of `C` is `void Function(Object?, [Object?])`, so ithe initialization of `t0Instance` should also succeed at run time. The initialization of `t1Instance` should also succeed, that's the straightforward case, and so is `t1Default`.

Reply:
May be then

> Check that if type `T0` is a function type obtained from a
> member signature with a covariant formal parameter and `T1` is a function
> type with formal parameters that are supertypes of those in `T0`, then `T0`
> is a subtype of `T1`.

?

Reply:
Maybe it just won't be a good match to talk about covariant parameters in the context of subtype tests at all? We don't have any types (function types, or indeed any type) with covariant parameter types, and hence it seems wrong to try to make a connection between subtype tests and tests about parameters that are covariant-by-declaration.

I think a natural thing to test would be (1) the static type of a method/setter tear-off ignores `covariant` (inherited as well as declared directly in the statically known type); `expectStaticType` should be able to handle this. (2) The dynamic type of a tear-off of a method/getter that has one or more parameters that are covariant-by-declaration has a top type as the parameter type for such parameters; tests like `e.foo is void Function(Object?, [Object?])` should be able to test this, because an actual run-time type of `void Function(T1, [T2])` would yield false for every pair of types `T1`, `T2` which is not top types. Finally, (3), we should test that a method/setter invocation and an invocation of a torn-off method/setter will actually throw if they have a parameter which is covariant-by-declaration, and the actual argument has a type which is OK for some supertype, but not a subtype of the actual parameter type of the function/method which is being called.

(OK, we can't tear off a setter, so we don't need to worry about the static or dynamic type of a torn off setter, but methods can do it, and the non-tear-off part of (3) can be done with setters as well).

We should be able to use a suitable receiver type resp. function type in order to enable testing of (3), but if there is an impossible case then we can always use a dynamic invocation.

Reply:
Please see https://github.com/dart-lang/co19/pull/2121